### ISSUE TITLE
Improper implementation of PHP Autoloader

### ISSUE THREAD

**Issue by bshaffer** [google/protobuf#3107]
The [PHP Autoloader](https://github.com/google/protobuf/blob/master/composer.json#L20) is including the class files at the start of each request instead of using proper PSR standards. This means if the `protobuf.so` C extension is used in tandem with this library, a fatal error is thrown due to the classes being redeclared.

If the classes are split into files based on their classnames, and `Classmap` or `PSR-0` is used for the autoloading instead of `files`, this will be resolved.

**Comment by stanley-cheung** [comment]
Here's the error if I load the protobuf via both composer and pecl

```
PHP Fatal error:  Cannot redeclare class Google\Protobuf\Internal\Descriptor in /home/stanleycheung/grpc/examples/php/vendor/google/protobuf/php/src/Google/Protobuf/descriptor.php on line 96
PHP Stack trace:
PHP   1. {main}() /home/stanleycheung/grpc/examples/php/route_guide/route_guide_client.php:0
PHP   2. require() /home/stanleycheung/grpc/examples/php/route_guide/route_guide_client.php:35
PHP   3. ComposerAutoloaderInitdcce3c6391bcbad0cee5c2b1d4bfac07::getLoader() /home/stanleycheung/grpc/examples/php/vendor/autoload.php:7
PHP   4. composerRequiredcce3c6391bcbad0cee5c2b1d4bfac07() /home/stanleycheung/grpc/examples/php/vendor/composer/autoload_real.php:56
PHP   5. require() /home/stanleycheung/grpc/examples/php/vendor/composer/autoload_real.php:66
```

**Comment by TeBoring** [comment]
I don't think the php implementation and c extension are supposed to be used together.
What's the use case?

**Comment by stanley-cheung** [comment]
The use case is that, if a user decides to install the protobuf C extension (for better performance), they usually would have "extension=protobuf.so" added to the `php.ini` file. So every time PHP is loaded, the protobuf C classes will be loaded.

Now let's say the user is also using some other composer packages. And thru dependency, the `google/protobuf` composer package is pulled in. That will create the error above.

So it's less about user deliberately using both at the same time. But as a consequence of using composer.

**Comment by bshaffer** [comment]
@TeBoring they aren't supposed to be used together, but also they shouldn't throw a PHP fatal error if they are accidentally used together. There's no reason for this, since PHP's autoloader is only activated when a class is missing.

One use case is if you have a user install a library requiring `grpc/grpc`, and they have the `grpc.so` extension installed, the library calls will work out of the box if `google/protobuf` is also required. However, if they'd like to optimize the protobuf layer, they can enable the `protobuf.so` extension at any time and are immediately making calls to the C library. This provides an easy default path and an easy upgrade path, without requiring every user to manage this themselves.

**Comment by bshaffer** [comment]
This was resolved in #3123
